src/wallet/wallet.cpp

```diff
@@ -3027,9 +3027,27 @@ int64_t CWallet::GetOldestKeyPoolTime()
     if (setKeyPool.empty())
         return GetTime();
 
-    // load oldest key from keypool, get time and return
     CKeyPool keypool;
     CWalletDB walletdb(strWalletFile);
+
+    if (IsHDEnabled() && CanSupportFeature(FEATURE_HD_SPLIT))
+    {
+        // if HD & HD Chain Split is enabled, response max(oldest-internal-key, oldest-external-key)
+        int64_t now = GetTime();
+        int64_t oldest_external = now, oldest_internal = now;
+
+        for(const int64_t& id : setKeyPool)
+        {
+            if (!walletdb.ReadPool(id, keypool))
+                throw runtime_error(std::string(__func__) + ": read failed");
+            if (keypool.fInternal && keypool.nTime < oldest_internal)
+                oldest_internal = keypool.nTime;
+            else if (!keypool.fInternal && keypool.nTime < oldest_external)
+                oldest_external = keypool.nTime;
+        }
+        return std::max(oldest_internal, oldest_external);
+    }
+    // load oldest key from keypool, get time and return
```

Comment:
I debugged that a little bit and it seems that this assumption (first item in setKeyPool is always oldest key) holds.

* The `std::set setKeyPool` (always ordered) contains `int64_t`. CKeyPool instances remember that nIndex (index of the setKeyPool int64.
* New CKeyPools will always insert `int64_t nEnd = *(--setKeyPool.end()) + 1;`.
* Returning keys will keep the date order because the returned will be inserted at the position where it has been taken out.

The only problem I can see is when someone overflows int64 by creating 2^63+1 keys. But, well,... yes.